packages/angular_devkit/build_angular/src/builders/browser-esbuild/compiler-plugin.ts

```diff
@@ -0,0 +1,411 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import type { CompilerHost } from '@angular/compiler-cli';
+import { transformAsync } from '@babel/core';
+import * as assert from 'assert';
+import type { OnStartResult, PartialMessage, PartialNote, Plugin, PluginBuild } from 'esbuild';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import ts from 'typescript';
+import { requiresLinking } from '../../babel/webpack-loader';
+import { loadEsmModule } from '../../utils/load-esm';
+import { BundleStylesheetOptions, bundleStylesheetData, bundleStylesheetFile } from './stylesheets';
+
+interface EmitFileResult {
+  content?: string;
+  map?: string;
+  dependencies: readonly string[];
+  hash?: Uint8Array;
+}
+type FileEmitter = (file: string) => Promise<EmitFileResult | undefined>;
+
+/**
+ * Converts TypeScript Diagnostic related information into an esbuild compatible note object.
+ * Related information is a subset of a full TypeScript Diagnostic and also used for diagnostic
+ * notes associated with the main Diagnostic.
+ * @param diagnostic The TypeScript diagnostic relative information to convert.
+ * @param host A TypeScript FormatDiagnosticsHost instance to use during conversion.
+ * @returns An esbuild diagnostic message as a PartialMessage object
+ */
+function convertTypeScriptDiagnosticInfo(
+  info: ts.DiagnosticRelatedInformation,
+  host: ts.FormatDiagnosticsHost,
+  textPrefix?: string,
+): PartialNote {
+  let text = ts.flattenDiagnosticMessageText(info.messageText, host.getNewLine());
+  if (textPrefix) {
+    text = textPrefix + text;
+  }
+
+  const note: PartialNote = { text };
+
+  if (info.file) {
+    note.location = {
+      file: info.file.fileName,
+      length: info.length,
+    };
+
+    if (info.start) {
+      const { line, character } = ts.getLineAndCharacterOfPosition(info.file, info.start);
+      note.location.line = line + 1;
+      note.location.column = character;
+      const { line: lastLine } = ts.getLineAndCharacterOfPosition(
+        info.file,
+        info.file.text.length - 1,
+      );
+      const lineStart = ts.getPositionOfLineAndCharacter(info.file, line, 0);
+      const lineEnd =
+        line < lastLine
+          ? ts.getPositionOfLineAndCharacter(info.file, line + 1, 0)
+          : info.file.text.length;
+      note.location.lineText = info.file.text.slice(lineStart, lineEnd).trimEnd();
+    }
+  }
+
+  return note;
+}
+
+/**
+ * Converts a TypeScript Diagnostic message into an esbuild compatible message object.
+ * @param diagnostic The TypeScript diagnostic to convert.
+ * @param host A TypeScript FormatDiagnosticsHost instance to use during conversion.
+ * @returns An esbuild diagnostic message as a PartialMessage object
+ */
+function convertTypeScriptDiagnostic(
+  diagnostic: ts.Diagnostic,
+  host: ts.FormatDiagnosticsHost,
+): PartialMessage {
+  let codePrefix = 'TS';
+  let code = `${diagnostic.code}`;
+  if (diagnostic.source === 'ngtsc') {
```

Comment:
It transforms the full, final string output of a diagnostic with a regexp (`TS-99` --> `NG`).  It's unfortunately not the best fit here since the string hasn't been constructed yet and the lower-level functions to only do the string replacement don't appear to be exported.